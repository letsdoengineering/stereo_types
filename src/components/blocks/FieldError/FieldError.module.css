.root {
    border: 2px solid var(--red);
    padding: 4px;
}

.children-wrapper {
    display: inline-flex;
    margin-bottom: 4px;
}
.error-message {
    font-size: smaller;
    display: flex;
    color: red;
}

