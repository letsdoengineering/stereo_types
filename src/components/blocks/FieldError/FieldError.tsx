import React from 'react'
import Text from '../../basics/Text/Text'

import * as fieldErrorStyles from './FieldError.module.css'

type FieldErrorPros = {
  children: React.ReactElement
  /** The message that should be displayed */
  errorMessage: string
  /** Flag to pass in error state */
  showError: boolean
  /** id must be provided, should be passing to the input child on the aria-describedby prop  */
  id: string
}

const FieldError: React.FC<FieldErrorPros> = ({ children, errorMessage, showError, id }) => {
  if (!showError) return children

  return (
    <div className={fieldErrorStyles.root}>
      <span className={fieldErrorStyles.childrenWrapper}>{children}</span>
      <Text id={id} weight='Bold' className={fieldErrorStyles.errorMessage}>
        {errorMessage}
      </Text>
    </div>
  )
}

export default FieldError
