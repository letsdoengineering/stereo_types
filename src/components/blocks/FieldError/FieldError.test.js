import * as React from 'react'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import { axe } from 'jest-axe'

import FieldError from './FieldError'

describe('FieldError', () => {
    it('displays error when showError is true', () => {
        render(
            <FieldError
                showError={true}
                errorMessage={
                    'You cannot travel to Norway, brown cheese is not a cheese'
                }
            >
                <p>I am a Child</p>
            </FieldError>
        )

        const errorMessage = screen.getByText(
            'You cannot travel to Norway, brown cheese is not a cheese'
        )
        expect(errorMessage).toBeInTheDocument()
    })

    it('only renders children when showError is false', () => {
        render(
            <FieldError
                showError={false}
                errorMessage={
                    'You cannot travel to Norway, brown cheese is not a cheese'
                }
            >
                <p>I am a Child</p>
            </FieldError>
        )

        const errorMessage = screen.queryByText(
            'You cannot travel to Norway, brown cheese is not a cheese'
        )
        expect(errorMessage).not.toBeInTheDocument()
    })

    it('has not changed the JSX layout of the page from previous snapshot', () => {
        const tree = renderer
            .create(
                <FieldError
                    showError={true}
                    errorMessage={'I am a error message'}
                    uniqueID='iamsuprisinglyunique'
                >
                    I am a Child Component
                </FieldError>
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders the page DOM that is accessible', async () => {
        const { container } = render(
            <FieldError showError={true} errorMessage={'I am a error message'}>
                <label>
                    label for input
                    <input type='text' />
                </label>
            </FieldError>
        )
        const results = await axe(container)

        expect(results).toHaveNoViolations()
    })
})
