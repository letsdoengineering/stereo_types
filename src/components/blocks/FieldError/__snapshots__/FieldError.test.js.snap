// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldError has not changed the JSX layout of the page from previous snapshot 1`] = `
<div
  className="root"
>
  <span
    className="childrenWrapper"
  >
    I am a Child Component
  </span>
  <span
    className="sizeS weightBold textGray errorMessage"
  >
    I am a error message
  </span>
</div>
`;
